src/dynamic_honey_badger/dynamic_honey_badger.rs

```diff
@@ -108,6 +215,154 @@ where
+            MessageContent::SignedVote(signed_vote) => self
+                .vote_counter
+                .add_pending_vote(sender_id, signed_vote)
+                .map(FaultLog::into),
+        }?;
+        let our_id = self.netinfo.our_id();
+        let all_remote_validators = self
+            .netinfo
+            .all_ids()
+            .filter(|&id| id != our_id)
+            .into_iter();
+        let all_remote_nodes = all_remote_validators.chain(self.nodes_being_added.iter());
+        let deferred_msgs = step.defer_messages(
+            &self.remote_epochs,
+            all_remote_nodes,
+            self.max_future_epochs as u64,
+        );
+        // Append the deferred messages onto the queues.
+        for (id, message) in deferred_msgs {
+            let epoch = message.dynamic_epoch();
+            if message.is_dynamic() {
+                self.outgoing_queue_dhb
+                    .entry((id, epoch.start_epoch))
+                    .and_modify(|e| e.push(message.clone()))
+                    .or_insert_with(|| vec![message.clone()]);
+            } else {
+                self.outgoing_queue_hb
+                    .entry((id, epoch))
+                    .and_modify(|e| e.push(message.clone()))
+                    .or_insert_with(|| vec![message.clone()]);
+            }
+        }
+        Ok(step)
+    }
+
+    /// Handles an epoch start announcement from a remote `DynamicHoneyBadger`.
+    fn handle_dynamic_epoch_started(
+        &mut self,
+        sender_id: &N,
+        epoch: DynamicEpoch,
+    ) -> Result<Step<C, N>> {
+        debug!(
+            "{:?}@{} received epoch {:?}@{:?}",
+            self.netinfo.our_id(),
+            self.start_epoch,
+            sender_id,
+            epoch,
+        );
+        let mut obsolete = false;
+        self.remote_epochs
+            .entry(sender_id.clone())
+            .and_modify(|e| {
+                if *e < epoch {
+                    // Increment the epoch.
+                    *e = epoch;
+                } else if *e >= epoch {
+                    obsolete = true;
+                }
+            }).or_insert(epoch);
+        if obsolete {
+            return Err(ErrorKind::ObsoleteDynamicEpochStarted.into());
+        }
+        let step = self.process_new_epoch(sender_id, epoch);
+        self.remove_messages_upto(sender_id, epoch);
+        step
+    }
+
+    /// Processes an announcement of a new epoch update received from a remote node.
+    fn process_new_epoch(&mut self, sender_id: &N, epoch: DynamicEpoch) -> Result<Step<C, N>> {
+        // Send any DHB messages for `epoch.start_epoch`.
+        let mut ready_messages = self
+            .outgoing_queue_dhb
+            .remove(&(sender_id.clone(), epoch.start_epoch))
+            .unwrap_or_else(|| vec![]);
+        // Send any HB messages for `epoch`.
+        ready_messages.extend(
+            self.outgoing_queue_hb
+                .remove(&(sender_id.clone(), epoch))
```

Comment:
Here we should probably include not only `epoch` but also the next `max_future_epochs` HB epochs?